### PR TITLE
Move function's arguments to registers

### PR DESCRIPTION
**Pull request by ruben-ayrapetyan**
Benchmark |                              Rss |                Perf
---- | ---- | ----
                                ./sunspider-1.0.2/3d-cube.js |    136 ->    132 (2.941%) | 1.88667 -> 1.90167 (-0.795%) | 
                    ./sunspider-1.0.2/access-binary-trees.js |     92 ->     84 (8.696%) | 1.51667 ->   1.01 (33.407%) | 
                        ./sunspider-1.0.2/access-fannkuch.js |     48 ->     48 (0.000%) | 5.00667 -> 4.75833 (4.960%) | 
                           ./sunspider-1.0.2/access-nbody.js |     68 ->     64 (5.882%) | 2.39333 ->   2.43 (-1.532%) | 
               ./sunspider-1.0.2/bitops-3bit-bits-in-byte.js |     40 ->     36 (10.000%) |  1.535 ->   1.25 (18.567%) | 
                    ./sunspider-1.0.2/bitops-bits-in-byte.js |     40 ->     36 (10.000%) | 2.01167 -> 1.73167 (13.919%) | 
                     ./sunspider-1.0.2/bitops-bitwise-and.js |     32 ->     32 (0.000%) | 1.88833 -> 1.80833 (4.237%) | 
                  ./sunspider-1.0.2/controlflow-recursive.js |    212 ->    216 (-1.887%) |   1.67 ->   0.79 (52.695%) | 
                            ./sunspider-1.0.2/crypto-sha1.js |    148 ->    144 (2.703%) | 8.47333 -> 9.03833 (-6.668%) | 
                      ./sunspider-1.0.2/date-format-xparb.js |    100 ->     92 (8.000%) | 1.22833 ->  1.295 (-5.428%) | 
                            ./sunspider-1.0.2/math-cordic.js |     44 ->     44 (0.000%) | 2.34333 -> 1.88333 (19.630%) | 
                      ./sunspider-1.0.2/math-partial-sums.js |     44 ->     44 (0.000%) | 1.44667 -> 1.06833 (26.152%) | 
                     ./sunspider-1.0.2/math-spectral-norm.js |     52 ->     48 (7.692%) | 1.61667 ->  1.085 (32.887%) | 
                           ./sunspider-1.0.2/string-fasta.js |     56 ->     56 (0.000%) |   3.09 -> 2.61333 (15.426%) | 
                                              Geometric mean |                          3.9506% |                         16.6639%

212fe8e (after rebase on current master):

                                                   Benchmark |                              Rss |                             Perf
---- | ---- | ----
                                ./sunspider-1.0.2/3d-cube.js |    136 ->    128 (5.882%) |  1.907 ->  1.963 (-2.937%) |                                                                                                                                                         
                    ./sunspider-1.0.2/access-binary-trees.js |     92 ->     84 (8.696%) | 1.5345 -> 1.0755 (29.912%) |                                                                                                                                                         
                        ./sunspider-1.0.2/access-fannkuch.js |     48 ->     48 (0.000%) |  5.104 -> 4.8395 (5.182%) |                                                                                                                                                          
                           ./sunspider-1.0.2/access-nbody.js |     68 ->     64 (5.882%) | 2.5045 ->  2.438 (2.655%) |                                                                                                                                                          
               ./sunspider-1.0.2/bitops-3bit-bits-in-byte.js |     40 ->     36 (10.000%) |   1.55 -> 1.3025 (15.968%) |                                                                                                                                                        
                    ./sunspider-1.0.2/bitops-bits-in-byte.js |     40 ->     36 (10.000%) | 2.0485 -> 1.8495 (9.714%) |                                                                                                                                                         
                     ./sunspider-1.0.2/bitops-bitwise-and.js |     32 ->     32 (0.000%) | 2.0025 ->  2.237 (-11.710%) |                                                                                                                                                        
                  ./sunspider-1.0.2/controlflow-recursive.js |    212 ->    216 (-1.887%) | 1.6865 -> 0.9075 (46.190%) |                                                                                                                                                        
                            ./sunspider-1.0.2/crypto-sha1.js |    148 ->    144 (2.703%) |  8.878 -> 8.6815 (2.213%) |                                                                                                                                                          
                      ./sunspider-1.0.2/date-format-xparb.js |    100 ->     92 (8.000%) | 1.2355 ->   1.31 (-6.030%) |                                                                                                                                                         
                            ./sunspider-1.0.2/math-cordic.js |     44 ->     44 (0.000%) |  2.406 -> 2.0335 (15.482%) |                                                                                                                                                         
                      ./sunspider-1.0.2/math-partial-sums.js |     44 ->     44 (0.000%) |  1.422 -> 1.3165 (7.419%) |                                                                                                                                                          
                     ./sunspider-1.0.2/math-spectral-norm.js |     52 ->     48 (7.692%) | 1.6365 ->  1.168 (28.628%) |                                                                                                                                                         
                           ./sunspider-1.0.2/string-fasta.js |     56 ->     56 (0.000%) |  3.188 ->  2.709 (15.025%) |                                                                                                                                                         
Geometric mean |                          4.1614% |                         12.6967%

### CODE REVIEW

File: jerry-core/parser/js/collections/linked-list.cpp [hunk 1]
```diff
@@ -29,67 +34,160 @@ do { \
...
-    printf ("Out of memory");
-    JERRY_UNREACHABLE ();
-  }
-  memset (list, 0, size);
-  linked_list_header* header = (linked_list_header *) list;
-  header->next = null_list;
-  header->element_size = element_size;
+  JERRY_ASSERT (list != null_list);
+
+  linked_list_header* header_p = (linked_list_header *) list;
```
Comment (egavrin) [inline jerry-core/parser/js/collections/linked-list.cpp]: ^_^ `linked_list_header *header_p`

**Comment by egavrin** [comment]
Good to me

File: jerry-core/parser/js/collections/linked-list.cpp [hunk 2]
```diff
@@ -17,9 +17,14 @@
...
-typedef struct linked_list_header
+typedef struct linked_list_chunk_header
 {
-  struct linked_list_header *next;
+  struct linked_list_chunk_header *next_p;
+} linked_list_chunk_header;
+
+typedef struct
+{
+  uint16_t list_length;
```
Comment (ruben-ayrapetyan) [inline jerry-core/parser/js/collections/linked-list.cpp]: Missing comments.

File: jerry-core/parser/js/collections/linked-list.cpp [hunk 3]
```diff
@@ -29,67 +34,160 @@ do { \
   JERRY_ASSERT (header); \
 } while (0);
 
-static size_t linked_list_block_size ()
+static size_t linked_list_block_size (bool is_first_chunk)
```
Comment (ruben-ayrapetyan) [inline jerry-core/parser/js/collections/linked-list.cpp]: Missing comments.

File: jerry-core/parser/js/collections/linked-list.cpp [hunk 4]
```diff
@@ -29,67 +34,160 @@ do { \
...
+  linked_list_chunk_header* chunk_header_p = (linked_list_chunk_header *) (header_p + 1u);
+
+  chunk_header_p->next_p = NULL;
+
   return list;
 }
 
+static linked_list_chunk_header *
+linked_list_append_new_chunk (linked_list_header *header_p,
+                              linked_list_chunk_header *last_chunk_header_p)
```
Comment (ruben-ayrapetyan) [inline jerry-core/parser/js/collections/linked-list.cpp]: Missing comments.

File: jerry-core/parser/js/opcodes-dumper.cpp [hunk 5]
```diff
@@ -2386,20 +2435,21 @@ dump_ret (void)
   serializer_dump_op_meta (jsp_dmp_create_op_meta_0 (VM_OP_RET));
 }
 
-void
+vm_instr_counter_t
 dump_reg_var_decl_for_rewrite (void)
```
Comment (ruben-ayrapetyan) [inline jerry-core/parser/js/opcodes-dumper.cpp]: Missing comments.

File: jerry-core/parser/js/opcodes-dumper.cpp [hunk 6]
```diff
@@ -2386,20 +2435,21 @@ dump_ret (void)
...
   dump_double_address (VM_OP_REG_VAR_DECL,
                        jsp_operand_t::make_unknown_operand (),
                        jsp_operand_t::make_unknown_operand ());
+
+  return oc;
 }
 
 void
-rewrite_reg_var_decl (void)
+rewrite_reg_var_decl (vm_instr_counter_t reg_var_decl_oc)
```
Comment (ruben-ayrapetyan) [inline jerry-core/parser/js/opcodes-dumper.cpp]: Missing comments.

File: jerry-core/parser/js/opcodes-dumper.cpp [hunk 7]
```diff
@@ -1182,6 +1241,17 @@ rewrite_function_end ()
   STACK_DROP (function_ends, 1);
 }
 
+void
+dumper_decrement_function_end_pos (void)
```
Comment (ruben-ayrapetyan) [inline jerry-core/parser/js/opcodes-dumper.cpp]: Missing comments.

File: jerry-core/vm/opcodes.h [hunk 8]
```diff
@@ -164,8 +164,12 @@ typedef enum : vm_idx_t
   OPCODE_SCOPE_CODE_FLAGS_STRICT                       = (1u << 0), /**< code is strict mode code */
   OPCODE_SCOPE_CODE_FLAGS_NOT_REF_ARGUMENTS_IDENTIFIER = (1u << 1), /**< code doesn't reference
                                                                      *   'arguments' identifier */
-  OPCODE_SCOPE_CODE_FLAGS_NOT_REF_EVAL_IDENTIFIER      = (1u << 2)  /**< code doesn't reference
+  OPCODE_SCOPE_CODE_FLAGS_NOT_REF_EVAL_IDENTIFIER      = (1u << 2), /**< code doesn't reference
                                                                      *   'eval' identifier */
+  OPCODE_SCOPE_CODE_FLAGS_ARGUMENTS_ON_REGISTERS       = (1u << 3), /**< function's arguments are moved to registers,
+                                                                     *   so should not be initialized in lexical
+                                                                     *   environment */
```
Comment (ruben-ayrapetyan) [inline jerry-core/vm/opcodes.h]:
```cpp
/**<
 * function's arguments are moved to registers,
 * so should be initialized in vm registers,
 * and not in lexical environment
 */
```

File: jerry-core/parser/js/collections/linked-list.cpp (hunk 1 above)
Comment (ruben-ayrapetyan) [inline jerry-core/parser/js/collections/linked-list.cpp]: Ok =)

File: jerry-core/ecma/operations/ecma-function-object.cpp [hunk 9]
```diff
@@ -214,13 +244,40 @@ ecma_op_create_function_object (ecma_collection_header_t *formal_params_collecti
...
+  bool is_no_lex_env = false;
+
+  vm_instr_counter_t instr_pos = first_instr_pos;
+  opcode_scope_code_flags_t scope_flags = vm_get_scope_flags (bytecode_header_p, instr_pos++);
+
+  if (scope_flags & OPCODE_SCOPE_CODE_FLAGS_STRICT)
+  {
+    is_strict_mode_code = true;
+  }
+  if ((scope_flags & OPCODE_SCOPE_CODE_FLAGS_NOT_REF_ARGUMENTS_IDENTIFIER)
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-function-object.cpp]: I would prefer newlines here because it feels else-ifs

**Comment by ruben-ayrapetyan** [comment]
@zherczeg, pull request was updated.

**Comment by zherczeg** [comment]
LGTM

**Pull request merged**